rfcs/20211025-augru/README.md

```diff
@@ -0,0 +1,1482 @@
+Proposal for GRU with Attentional Update in DNNL
+================================================
+
+## 1. Introduction
+
+In addition to the regular GRU there is a notion of GRU with attentional update
+gate (AUGRU) that can help to model interest evolving in Click-Through Rate
+Prediction [[#1]][1].
+
+### 1.1. DIEN model
+
+DIEN model is a novel model used for Click-Through Rate Prediction [[#1]][1].
+It consists of Behavior Layer, Interest Extractor Layer and
+Interest Evolving Layer. In this proposal only Interest Evolving Layer is
+covered, because this is a new part which is not covered by oneDNN.
+Interest Evolving Layer consists of an attention mechanism and AUGRU:
+
+![InterestEvolvingLayer](diagram-iel.png)
+
+
+Attention mechanism is implemented as a fully-connected network (sequence of
+Matmul operations and a Softmax), hence this part is already covered by oneDNN.
+
+### 1.2. AUGRU
+
+AUGRU is defined by the following formula (the difference from the
+regular GRU is highlighted in red):
+
+![AUGRU-LBR](formula-augru-lbr.png)
+
+
+> Warning
+>
+> Attention is passed not as part of source layer as it is done in Neural Machine
+> Translation (NMT) models [[#3]][3], but as a separate memory which is
+> applied to intermediate results (to gate U). This difference makes it impossible
+> to apply attention on application side and pass the result to oneDNN RNN
+> primitive.
+
+
+## 2. Proposal
+
+AUGRU can't be implemented by existing oneDNN RNN primitives and it takes a big
+part of time of overall model execution, so adding it to the library will bring
+performance speed-up to the models which use AUGRU.
+
+There are at least 2 options on how we can add AUGRU support into oneDNN:
+ - a. Keeping API as is, reuse `weights_peephole_desc` for `attention_desc`;
+ - b. Introducing `rnn_desc_v2_t` with all necessary fields for AUGRU, providing
+ more flexibility to the users.
+
```

Comment:
maybe explicitely put the c. option you mention later: a combination of a and b with rnn_desc_v2_t that would rename weights_peephole_desc to make peephole and attention descs aliasing explicit. Either with pure naming (e.g. `lstm_peephole_augru_attention`) or using a union.

Reply:
done, thanks!